File: packages/browseros-agent/apps/eval/scripts/analyze-webbench.ts
/**
 * Analyze WebBench results across 4 agents to stratify tasks by pass count.
 * Usage: bun apps/eval/scripts/analyze-webbench.ts
 */
import { parse } from 'csv-parse/sync'

const dataDir = 'apps/eval/data/webbench'

interface AgentConfig {
  file: string
  evalCol: string
  name: string
}

const agents: AgentConfig[] = [
  { file: 'anthropicfinal.csv', evalCol: 'Anthropic_Eval', name: 'Anthropic' },
  { file: 'skyvern2.0final.csv', evalCol: 'Skyvern2.0Eval', name: 'Skyvern' },
  { file: 'openaicuafinal.csv', evalCol: 'CUAEval', name: 'OpenAI CUA' },
  { file: 'browserusefinal.csv', evalCol: 'BUEval', name: 'BrowserUse' },
]

type Row = Record<string, string>

// Parse each agent's results
const agentResults = new Map<
  string,
  Map<
    number,
    {
      eval: string
      difficulty: string
      category: string
      task: string
      url: string
    }
  >
>()

for (const agent of agents) {
  const text = await Bun.file(`${dataDir}/${agent.file}`).text()
  const rows: Row[] = parse(text, {
    columns: true,
    skip_empty_lines: true,
    relax_column_count: true,
  })
  const results = new Map<
    number,
    {
      eval: string
      difficulty: string
      category: string
      task: string
      url: string
    }
  >()
  for (const row of rows) {
    const id = parseInt(row.ID, 10)
    if (Number.isNaN(id)) continue
    results.set(id, {
      eval: row[agent.evalCol] || '',
      difficulty: row.Difficulty || '',
      category: row.Category || '',
      task: row.Task || '',
      url: row['Starting URL'] || '',
    })
  }
  agentResults.set(agent.name, results)
  console.log(`${agent.name}: ${results.size} tasks loaded`)
}

// Find common task IDs (present in all 4 agents)
const allIds = new Set<number>()
for (const [, results] of agentResults) {
  for (const id of results.keys()) allIds.add(id)
}

// Build pass count per task
interface TaskStats {
  id: number
  passCount: number
  difficulty: string
  category: string
  task: string
  url: string
  agents: Record<string, string>
}

const taskStats: TaskStats[] = []
const _fullAgentNames = agents.map((a) => a.name)

for (const id of allIds) {
  let passCount = 0
  let _presentCount = 0
  const agentEvals: Record<string, string> = {}
  let difficulty = ''
  let category = ''
  let task = ''
  let url = ''

  for (const agent of agents) {
    const result = agentResults.get(agent.name)?.get(id)
    if (result) {
      _presentCount++
      const isSuccess = result.eval?.toLowerCase().includes('success')
      if (isSuccess) passCount++
      agentEvals[agent.name] = isSuccess ? 'PASS' : 'FAIL'
      if (!difficulty) difficulty = result.difficulty
      if (!category) category = result.category
      if (!task) task = result.task
      if (!url) url = result.url
    } else {
      agentEvals[agent.name] = 'N/A'
    }
  }

  taskStats.push({
    id,
    passCount,
    difficulty,
    category,
    task,
    url,
    agents: agentEvals,
  })
}

// Group by pass count
const byPassCount: Record<number, TaskStats[]> = {
  0: [],
  1: [],
  2: [],
  3: [],
  4: [],
}
for (const t of taskStats) {
  byPassCount[t.passCount].push(t)
}

console.log('\n═══════════════════════════════════════════════════')
console.log('TASKS BY PASS COUNT (how many agents succeeded)')
console.log('═══════════════════════════════════════════════════\n')

for (let pc = 0; pc <= 4; pc++) {
  const tasks = byPassCount[pc]
  const label =
    pc === 0 ? '0/4 (ALL FAIL)' : pc === 4 ? '4/4 (ALL PASS)' : `${pc}/4`
  console.log(`${label}: ${tasks.length} tasks`)

  // Breakdown by difficulty
  const easy = tasks.filter((t) => t.difficulty === 'easy').length
  const hard = tasks.filter((t) => t.difficulty === 'hard').length
  console.log(`  easy: ${easy}, hard: ${hard}`)

  // Breakdown by category
  const byCat: Record<string, number> = {}
  for (const t of tasks) {
    byCat[t.category] = (byCat[t.category] || 0) + 1
  }
  console.log(
    `  categories: ${Object.entries(byCat)
      .sort((a, b) => b[1] - a[1])
      .map(([c, n]) => `${c}(${n})`)
      .join(', ')}`,
  )
  console.log()
}

// Now handle BrowserUse only having 658 tasks — let's also do a 3-agent view (Anthropic, Skyvern, OpenAI)
console.log('\n═══════════════════════════════════════════════════')
console.log('3-AGENT VIEW (Anthropic + Skyvern + OpenAI CUA)')
console.log('(BrowserUse only has 658 tasks, so this is more complete)')
console.log('═══════════════════════════════════════════════════\n')

const threeAgents = ['Anthropic', 'Skyvern', 'OpenAI CUA']
const byPassCount3: Record<number, TaskStats[]> = { 0: [], 1: [], 2: [], 3: [] }

for (const t of taskStats) {
  let pc3 = 0
  let allPresent = true
  for (const a of threeAgents) {
    if (t.agents[a] === 'N/A') {
      allPresent = false
      break
    }
    if (t.agents[a] === 'PASS') pc3++
  }
  if (!allPresent) continue
  if (!byPassCount3[pc3]) byPassCount3[pc3] = []
  byPassCount3[pc3].push(t)
}

let total3 = 0
for (let pc = 0; pc <= 3; pc++) {
  const tasks = byPassCount3[pc]
  total3 += tasks.length
  const label =
    pc === 0 ? '0/3 (ALL FAIL)' : pc === 3 ? '3/3 (ALL PASS)' : `${pc}/3`
  console.log(`${label}: ${tasks.length} tasks`)

  const easy = tasks.filter((t) => t.difficulty === 'easy').length
  const hard = tasks.filter((t) => t.difficulty === 'hard').length
  console.log(`  easy: ${easy}, hard: ${hard}`)

  const byCat: Record<string, number> = {}
  for (const t of tasks) {
    byCat[t.category] = (byCat[t.category] || 0) + 1
  }
  console.log(
    `  categories: ${Object.entries(byCat)
      .sort((a, b) => b[1] - a[1])
      .map(([c, n]) => `${c}(${n})`)
      .join(', ')}`,
  )

  // Show unique websites count
  const uniqueUrls = new Set(tasks.map((t) => t.url))
  console.log(`  unique websites: ${uniqueUrls.size}`)
  console.log()
}
console.log(`Total tasks in 3-agent intersection: ${total3}`)

// Quick sample of 0/3 tasks (hardest)
console.log('\n── Sample 0/3 (all fail) tasks ──')
byPassCount3[0].slice(0, 5).forEach((t) => {
  console.log(`  [${t.id}] [${t.difficulty}] [${t.category}] ${t.url}`)
  console.log(`    ${t.task.slice(0, 150)}`)
})

console.log('\n── Sample 1/3 tasks ──')
byPassCount3[1].slice(0, 5).forEach((t) => {
  console.log(`  [${t.id}] [${t.difficulty}] [${t.category}] ${t.url}`)
  console.log(`    ${t.task.slice(0, 150)}`)
})

File: packages/browseros-agent/apps/eval/scripts/analyze-webbench.py
"""
Analyze WebBench results across ALL 8 agents to stratify tasks by pass count.
Usage: python3 apps/eval/scripts/analyze-webbench.py
"""
import csv
import os
from collections import defaultdict

DATA_DIR = "apps/eval/data/webbench"

AGENTS = [
    {"file": "anthropicfinal.csv", "eval_col": "Anthropic_Eval", "name": "Anthropic CUA"},
    {"file": "skyvern2.0final.csv", "eval_col": "Skyvern2.0Eval", "name": "Skyvern 2.0"},
    {"file": "skyvern2.0browserbasefinal.csv", "eval_col": "Browserbase_SkyvernEval", "name": "Skyvern BB"},
    {"file": "openaicuafinal.csv", "eval_col": "CUAEval", "name": "OpenAI CUA"},
    {"file": "browserusefinal.csv", "eval_col": "BUEval", "name": "BrowserUse"},
    {"file": "convergencehitlfinal.csv", "eval_col": "convergence_hitl_eval", "name": "Convergence"},
    {"file": "operatorhitlfinal.csv", "eval_col": "operator_hitl_eval", "name": "Operator"},
    {"file": "rtrvrfinal.csv", "eval_col": "Human Label", "name": "RTRVR"},
]


def load_agent(agent):
    path = os.path.join(DATA_DIR, agent["file"])
    results = {}
    with open(path, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            try:
                task_id = int(row["ID"])
            except (ValueError, KeyError):
                continue
            eval_val = row.get(agent["eval_col"], "")
            results[task_id] = {
                "eval": eval_val,
                "difficulty": row.get("Difficulty", ""),
                "category": row.get("Category", ""),
                "task": row.get("Task", ""),
                "url": row.get("Starting URL", ""),
            }
    return results


# Load all agents
print("Loading agents...")
agent_results = {}
for agent in AGENTS:
    data = load_agent(agent)
    agent_results[agent["name"]] = data
    print(f"  {agent['name']}: {len(data)} tasks")

# ─── INDIVIDUAL AGENT STATS ──────────────────────────────────────────
print("\n" + "=" * 70)
print("INDIVIDUAL AGENT PASS RATES")
print("=" * 70)

for agent in AGENTS:
    name = agent["name"]
    data = agent_results[name]
    total = len(data)
    passed = sum(1 for r in data.values() if r["eval"] and "success" in r["eval"].lower())
    easy_total = sum(1 for r in data.values() if r["difficulty"] == "easy")
    easy_pass = sum(1 for r in data.values() if r["difficulty"] == "easy" and r["eval"] and "success" in r["eval"].lower())
    hard_total = sum(1 for r in data.values() if r["difficulty"] == "hard")
    hard_pass = sum(1 for r in data.values() if r["difficulty"] == "hard" and r["eval"] and "success" in r["eval"].lower())
    print(f"\n{name}: {passed}/{total} = {passed/total*100:.1f}%")
    if easy_total:
        print(f"  easy: {easy_pass}/{easy_total} = {easy_pass/easy_total*100:.1f}%")
    if hard_total:
        print(f"  hard: {hard_pass}/{hard_total} = {hard_pass/hard_total*100:.1f}%")

# ─── FULL-COVERAGE AGENTS (2452 tasks each) ──────────────────────────
# Anthropic CUA, Skyvern 2.0, Skyvern BB, OpenAI CUA
full_agents = ["Anthropic CUA", "Skyvern 2.0", "Skyvern BB", "OpenAI CUA"]

print("\n" + "=" * 70)
print(f"4 FULL-COVERAGE AGENTS: {', '.join(full_agents)}")
print("(each has ~2452 tasks)")
print("=" * 70)

# Collect IDs present in ALL 4 full agents
all_ids = None
for name in full_agents:
    ids = set(agent_results[name].keys())
    all_ids = ids if all_ids is None else all_ids & ids

print(f"Tasks in intersection: {len(all_ids)}")

by_pass = defaultdict(list)
for tid in sorted(all_ids):
    pass_count = 0
    info = {}
    agent_evals = {}
    for name in full_agents:
        r = agent_results[name][tid]
        is_success = "success" in r["eval"].lower() if r["eval"] else False
        if is_success:
            pass_count += 1
        agent_evals[name] = "PASS" if is_success else "FAIL"
        if not info:
            info = r
    by_pass[pass_count].append({
        "id": tid, "pass_count": pass_count,
        "difficulty": info["difficulty"], "category": info["category"],
        "task": info["task"], "url": info["url"], "agents": agent_evals,
    })

for pc in range(5):
    tasks = by_pass[pc]
    label = {0: "0/4 (ALL FAIL)", 4: "4/4 (ALL PASS)"}.get(pc, f"{pc}/4")
    easy = sum(1 for t in tasks if t["difficulty"] == "easy")
    hard = sum(1 for t in tasks if t["difficulty"] == "hard")
    cats = defaultdict(int)
    for t in tasks:
        cats[t["category"]] += 1
    urls = len(set(t["url"] for t in tasks))
    cat_str = ", ".join(f"{c}({n})" for c, n in sorted(cats.items(), key=lambda x: -x[1]))
    print(f"\n{label}: {len(tasks)} tasks")
    print(f"  easy: {easy}, hard: {hard}")
    print(f"  categories: {cat_str}")
    print(f"  unique websites: {urls}")

# ─── NOW ALSO CHECK: how many 0/4 tasks require login? ───────────────
print("\n" + "=" * 70)
print("0/4 TASKS: LOGIN vs NO-LOGIN breakdown")
print("=" * 70)

login_keywords = ["log in", "login", "sign in", "signin", "your account", "your profile",
                   "your wishlist", "your order", "your cart", "your dashboard", "your settings",
                   "your subscription", "your inbox", "your message", "your review",
                   "send a message", "post a comment", "write a review", "submit a",
                   "publish", "upload"]
zero_pass = by_pass[0]
login_tasks = []
no_login_tasks = []
for t in zero_pass:
    task_lower = t["task"].lower()
    needs_login = any(kw in task_lower for kw in login_keywords)
    if needs_login:
        login_tasks.append(t)
    else:
        no_login_tasks.append(t)

print(f"  Likely needs login: {len(login_tasks)}")
print(f"  Possibly no login:  {len(no_login_tasks)}")

print(f"\n  No-login 0/4 tasks by category:")
cats = defaultdict(int)
for t in no_login_tasks:
    cats[t["category"]] += 1
cat_str = ", ".join(f"{c}({n})" for c, n in sorted(cats.items(), key=lambda x: -x[1]))
print(f"    {cat_str}")

print(f"\n  Sample no-login 0/4 tasks:")
for t in no_login_tasks[:10]:
    print(f"    [{t['id']}] [{t['difficulty']}] [{t['category']}] {t['url']}")
    print(f"      {t['task'][:180]}")

# ─── ALSO INCLUDE THE HITL AGENTS (smaller overlap) ──────────────────
hitl_agents = ["Convergence", "Operator", "RTRVR"]
print("\n" + "=" * 70)
print(f"HITL AGENTS: {', '.join(hitl_agents)}")
print("=" * 70)

for name in hitl_agents:
    data = agent_results[name]
    total = len(data)
    passed = sum(1 for r in data.values() if r["eval"] and "success" in r["eval"].lower())
    print(f"  {name}: {passed}/{total} = {passed/total*100:.1f}%")

# See how HITL agents do on the same tasks as the 4 full agents
hitl_ids = None
for name in hitl_agents:
    ids = set(agent_results[name].keys())
    hitl_ids = ids if hitl_ids is None else hitl_ids & ids

common_hitl = all_ids & hitl_ids if hitl_ids else set()
print(f"\n  Tasks in common (all 7 agents): {len(common_hitl)}")

if common_hitl:
    by_pass_7 = defaultdict(list)
    all_7 = full_agents + hitl_agents
    for tid in sorted(common_hitl):
        pass_count = 0
        info = {}
        for name in all_7:
            r = agent_results[name].get(tid)
            if r:
                is_success = "success" in r["eval"].lower() if r["eval"] else False
                if is_success:
                    pass_count += 1
                if not info:
                    info = r
        by_pass_7[pass_count].append({"id": tid, **info})

    print("\n  7-AGENT PASS COUNT (on common subset):")
    for pc in range(8):
        if by_pass_7[pc]:
            print(f"    {pc}/7: {len(by_pass_7[pc])} tasks")

# ─── SUMMARY TABLE ───────────────────────────────────────────────────
print("\n" + "=" * 70)
print("SUMMARY FOR DATASET BUILDING")
print("=" * 70)
print(f"""
Pool sizes (4 full-coverage agents):
  0/4 (all fail):  {len(by_pass[0]):>4}  (login-required: ~{len(login_tasks)}, no-login: ~{len(no_login_tasks)})
  1/4:             {len(by_pass[1]):>4}
  2/4:             {len(by_pass[2]):>4}
  3/4:             {len(by_pass[3]):>4}
  4/4 (all pass):  {len(by_pass[4]):>4}
  ─────────────────────
  Total:           {sum(len(v) for v in by_pass.values()):>4}
""")
